### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by JMurphyWeb**
@mk4111 feel free to comment/question anything that you want to go over. It took a while to figure out how to get started but im feeling a bit more comfortable with it now.

This is as far as i got. I left a couple of comments in the server/auth.js file which is where i registered the strategy using hapi-auth-cookie module/plugin/scheme!?

At the moment the strategy's mode is set to true (line 8) which means it will be used for every route unless specified (instructions on how to do so are in auth.js). For development we can always change this to false so that no routes will be authenticated.

In order to get an idea of how the flow works, i have left a dummy route in there `/set` which will set your cookie {userId: 0}, you should then be allowed to visit the route `/`without being redirected to `/login` **assuming you have at least 1 active user in your db**

PS did i do the github base-branch thing right?

#67 contains my remaining todos for this PR

### CODE REVIEW

**Comment by JMurphyWeb** [comment]
previous commit references #67

**Comment by JMurphyWeb** [comment]
@mk4111 i've added a check in auth.js for the user_type of the incoming user, and added that user_type as a `scope` attribute to the request.auth object. 

Each route can define which users can visit it by adding the following object to the route config.auth:
`scope: 'admin'` or `scope: ['admin', 'primary']`

**Comment by JMurphyWeb** [comment]
@nelsonic I have added auth and it would be great to have a bit of feedback on my implementation. 

All our old tests are failing on this branch as the requests are not authenticated. Should i manually add a cookie to each request? is it ok to create a never-expiring cookie and store it as a environment variable to be used for testing?  Other options i considered were: 

- Generating a new cookie each time the tests run with iron.seal (this would mean wrapping all the tests which require auth in a callback, which seems cumbersome)
- "mock" a login by hitting the /login route at the beginning of each test, which again seems pretty cumbersome.

I am assigning you but this is definitely not ready to be merged, please let me know if that is not the correct way to go about it.

**Comment by JMurphyWeb** [comment]
@mk4111 should the auth plugin update the user's last login if successful? Seems like a sensible place to perform an update but it would mean the plugin is doing more than just auth. What do you think?

File: package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "boom": "^3.2.2",
     "handlebars": "^4.0.5",
     "hapi": "^14.2.0",
+    "hapi-auth-cookie": "^6.1.1",
```
Comment (nelsonic) [inline package.json]: we can do better... https://github.com/dwyl/hapi-auth-jwt2#want-to-sendstore-your-jwt-in-a-cookie

File: server/handlers.js [hunk 2]
```diff
@@ -7,25 +7,44 @@ const hash = new Hashids(process.env.HASHID_KEY);
 
 const handlers = {};
 
-const cookieConfig = {
-  ttl: 7 * 24 * 60 * 60 * 1000,
-  isSecure: false,
-  path: '/'
+handlers.serveView = viewName => (request, reply) => {
```
Comment (nelsonic) [inline server/handlers.js]: un-used. please remove.

File: server/routes.js [hunk 3]
```diff
@@ -1,16 +1,43 @@
 const validate = require('./joi.js');
 const handlers = require('./handlers.js');
 
+// TODO: check routes
```
Comment (nelsonic) [inline server/routes.js]: ❓

File: server/routes.js [hunk 4]
```diff
@@ -1,16 +1,43 @@
...
-    handler: (request, reply) => reply('Hello World')
+    handler: handlers.checkUser, // replace with handlers.serveSpecificFile('dashboard')
+    config: { auth: false }
+  },
+  {
+    method: 'GET',
+    path: '/get',
+    handler: (request, reply) => {
+      // console.log(request.auth);
+      reply('Hello World');
```
Comment (nelsonic) [inline server/routes.js]: please _remove_ this.

File: server/routes.js [hunk 5]
```diff
@@ -1,16 +1,43 @@
...
+    },
+    config: {
+      auth: false // {
+      //   scope: 'primary'
+      // }
+    }
+  },
+  {
+    method: 'GET',
+    path: '/set',
```
Comment (nelsonic) [inline server/routes.js]: _what is this doing?_

File: server/routes.js [hunk 6]
```diff
@@ -1,16 +1,43 @@
...
+    },
+    config: {
+      auth: false
+    }
   },
   {
     method: 'GET',
     path: '/people/add',
-    handler: handlers.serveSpecificFile('add-user')
+    handler: handlers.serveView('add-user')
```
Comment (nelsonic) [inline server/routes.js]: are these views _static_ ?

File: server/routes.js [hunk 7]
```diff
@@ -34,17 +61,44 @@ const routes = [
     }
   },
   {
+    method: 'GET',
+    path: '/login',
+    handler: handlers.serveView('login'),
```
Comment (nelsonic) [inline server/routes.js]: We _literally_ have a ***fully-working-example*** of Login: https://github.com/dwyl/hapi-login ...

Comment (iteles) [inline server/routes.js]: Which we in fact discussed should be added to the relevant issue as part of the discussion for this feature. Is there a reason you chose not to use this @jmurphyweb? Please add it to the issue (note: this is a good example of where it is advantageous to write out your thought process on how you will tackle an issue before implementation)

File: test/auth.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
+const tape = require('tape');
+const client = require('redis').createClient();
+const server = require('../server/server.js');
+const Iron = require('iron');
+const mockData = require('./mock-data.js');
+
+require('env2')('config.env');
```
Comment (nelsonic) [inline test/auth.js]: `require('env2')` _once_ in the "_main_" file (`../server/server.js`)

File: test/auth.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
+const tape = require('tape');
+const client = require('redis').createClient();
+const server = require('../server/server.js');
+const Iron = require('iron');
```
Comment (nelsonic) [inline test/auth.js]: `iron`...? why?

File: test/auth.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
+const tape = require('tape');
+const client = require('redis').createClient();
```
Comment (nelsonic) [inline test/auth.js]: thought we were _removing_ direct calls to the Redis DB from our tests...?

File: server/handlers.js (hunk 2 above)
Comment (JMurphyWeb) [inline server/handlers.js]: This is all deleted in the current refactor tests branch.

File: server/routes.js (hunk 4 above)
Comment (JMurphyWeb) [inline server/routes.js]: This is all deleted in the current refactor tests branch.

File: server/routes.js (hunk 5 above)
Comment (JMurphyWeb) [inline server/routes.js]: This is all deleted in the current refactor tests branch.

File: server/routes.js (hunk 6 above)
Comment (JMurphyWeb) [inline server/routes.js]: I don't think so. They are views built built from handlebars.

File: test/auth.js (hunk 8 above)
Comment (JMurphyWeb) [inline test/auth.js]: Ok great, thanks

File: test/auth.js (hunk 9 above)
Comment (JMurphyWeb) [inline test/auth.js]: As far as i am aware, hapi auth cookie uses iron. I have removed it from the testing files where it wasn't used. Is there a problem with using Iron in general?

File: test/auth.js (hunk 10 above)
Comment (JMurphyWeb) [inline test/auth.js]: That is what i was working on yesterday. It is on refactor tests branch which is yet to be PRd.

File: server/routes.js (hunk 7 above)
Comment (JMurphyWeb) [inline server/routes.js]: @iteles I will respond in #45 

Just to clarify this PR consists of a combination of my and @mk4111s code, and we looked into the plugin together in week 1.

**Comment by JMurphyWeb** [comment]
@nelsonic thanks for the feedback. Sorry if i was not entirely clear in my @ you at the top, I mainly wanted a check of the overall implementation of auth. There was a good bit of code left in there which was either me trying to work something out or left as an example to Marie to see how it worked. 

As I have commented most of this code is removed from the refined branch refactor-tests.  I will create a PR with a more appropriate name and a description later.

File: server/routes.js (hunk 3 above)
Comment (JMurphyWeb) [inline server/routes.js]: This is deleted in the current refactor tests branch.

**Comment by nelsonic** [comment]
@JMurphyWeb if this PR is still required, please merge master into the branch, fix merge conflicts then assign to me. thanks! 👍

**Pull request closed without merging**